harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,347 @@
+            )
+        return mpu
+
+    @property
+    def mpu(self) -> det_ds.ModelParallelUnit:
+        return self._mpu
+
+    def wrap_model_engine(self, model: torch.nn.Module) -> torch.nn.Module:
+        """Returns a wrapped model engine.
+
+        In the background, we perform checks to properly handle pipeline parallelism if
+        the model engine is a PipelineEngine.  We also recompute batch sizes to match
+        the deepspeed config.
+        """
+
+        if self.env.managed_training:
+            model = model.to(self.device)
+
+        # Pipeline parallel model engine has its own MPU that we will use here.
+        if isinstance(model, deepspeed.PipelineEngine):
+            self._use_pipeline_parallel = True
+            self._mpu = det_ds.DeepSpeedMPU(model.mpu)
+
+        recompute_batch_size = False
+
+        # Check to make sure that Determined's global batch size matches the model engine's.
+        # If not, overwrite Determined's batch size.
+        if self.get_global_batch_size() != model.train_batch_size():
+            logging.warning(
+                f"Setting global batch size to {model.train_batch_size()} to match the "
+                "deepspeed config.  To prevent this from happening, you can call "
+                "self.context.overwrite_deepspeed_config(base_ds_config) to get a consistent"
+                "deepspeed config dict which you can pass to the config field when calling"
+                "deepspeed.initialize to build the model engine."
+            )
+            self.env._global_batch_size = model.train_batch_size()
+            recompute_batch_size = True
+
+        if self._train_micro_batch_size_per_gpu is None:
+            self._train_micro_batch_size_per_gpu = int(model.train_micro_batch_size_per_gpu())
+            recompute_batch_size = True
+        # If multiple model engines are wrapped, we will make sure that they have the same
+        # micro batch size.
+        assert (
+            self._train_micro_batch_size_per_gpu == model.train_micro_batch_size_per_gpu()
+        ), "micro batch size do not match across DeepSpeed model engines."
+
+        if recompute_batch_size:
+            (
+                self.env._per_slot_batch_size,
+                self.env._global_batch_size,
+            ) = self._calculate_batch_sizes()
+
+        self.models.append(model)
+        return model
+
+    def disable_auto_grad_accumulation(self) -> None:
+        """
+        Prevent the DeepSpeedTrialController from automatically calling train_batch multiple times
+        to process enough micro batches to meet the per slot batch size.  Thus, the user is
+        responsible for manually training on enough micro batches in train_batch to meet the
+        expected per slot batch size.
+        """
+        self._manual_grad_accumulation = True
+
+    def disable_dataset_reproducibility_checks(self) -> None:
+        self._data_repro_checks_disabled = True
```

Comment:
side note: do you think we should promote `disable_dataset_reproducibility_checks` to non-experimental in pytorchtrial too?  (not as part of this change, I'm asking if you think the API is "right"... I don't know of a single user who has actually used it to date)

Reply:
note: if you don't think it's right yet then we need to update the error message to either say `context.experimental.disable_...` or `context.disable_...`

Reply:
i think we should promote `disable_dataset_reproducibility_checks` to non-experimental with support for native dtrain because i think it will be more likely that we'll have users familiar with the native DistributedSampler or will just use the distributed dataloaders available in the original code they have.